crates/bevy_asset/src/processor/mod.rs

```diff
@@ -227,89 +225,129 @@ impl AssetProcessor {
-            source_receivers.push(Box::pin(source_receiver));
-        }
-        let mut all_receiver = select_all(source_receivers);
-
-        // We want to await an entry in the stream, but once we have one, we want to process all the
-        // events in the channels before sending the "finished processing" state. So await the next
-        // item, then keep getting the next entry in the stream until we have to sleep.
-        let mut next = None;
-        while let Some(mut item) = {
-            if next.is_none() {
-                next = Some(all_receiver.next());
-            }
-            next.take().unwrap()
+            let processor = self.clone();
+            let sender = sender.clone();
+            IoTaskPool::get()
+                .spawn(async move {
+                    while let Ok(event) = receiver.recv().await {
+                        let Ok(source) = processor.get_source(source_id.clone()) else {
+                            return;
+                        };
+                        processor
+                            .handle_asset_source_event(source, event, &sender)
+                            .await;
+                    }
+                })
+                .detach();
         }
-        .await
-        {
-            self.set_state(ProcessorState::Processing).await;
+    }
 
-            loop {
-                let (source_id, event) = item;
-                self.handle_asset_source_event(self.data.sources.get(source_id).unwrap(), event)
-                    .await;
-                let mut next_next = all_receiver.next();
-                item = match poll_once(&mut next_next).await {
-                    None => {
-                        next = Some(next_next);
-                        break;
+    /// Executes all tasks that come through `receiver`, and updates the processor's overall state
+    /// based on task starts and ends.
+    async fn execute_processing_tasks(
+        &self,
+        sender: &async_channel::Sender<ProcessorTaskEvent>,
+        receiver: &async_channel::Receiver<ProcessorTaskEvent>,
+    ) {
+        // If there aren't any tasks in the channel the first time around, we should immediately go
+        // to the finished state (otherwise we'd be sitting around stuck in the `Initialized`
+        // state).
+        if receiver.is_empty() {
+            self.set_state(ProcessorState::Finished).await;
+        }
+        // Note: only this method sends `ProcessorTaskEvent::Finished` (to itself). We could have
+        // two channels, one for finished messages, and one for starting tasks, but this requires
+        // selecting across them, which the macro is feature-gated behind `std`.
+        // TODO: This task currently holds both the sender and receiver, meaning the channel never
+        // "naturally" closes, so this task never terminates. It would be nice if this task
+        // naturally terminated when the asset source event receivers close.
```

Comment:
Surely it must be possible to select in no_std? Otherwise it's quite limiting for bevy_asset to support it :D